src/plugins/intel_cpu/src/nodes/subgraph.cpp

```diff
@@ -43,65 +46,218 @@ namespace intel_cpu {
+        auto broadcast_merge = [](VectorDims& dst, const VectorDims& src) {
+            // Ranks are both static.
+            auto dst_rank = dst.size();
+            auto src_rank = src.size();
+            const auto new_rank = std::max(dst_rank, src_rank);
+            dst.insert(dst.begin(), new_rank - dst_rank, 1);
+            for (size_t i = 0; i < new_rank; i++) {
+                auto srci = i < (new_rank - src_rank) ? 1 : src[i - (new_rank - src_rank)];
+                if (dst[i] != srci && srci != Shape::UNDEFINED_DIM) {
+                    if (dst[i] == 1 || dst[i] == Shape::UNDEFINED_DIM)
+                        dst[i] = srci;
+                }
+            }
+        };
+
+        const size_t out_size = m_body->get_output_size();
+        if (out_size == 1) {
+            VectorDims masterShape;
+            for (size_t i = 0; i < input_shapes.size(); i++) {
+                if (i == 0)
+                    masterShape = input_shapes[i];
+                else
+                    broadcast_merge(masterShape, input_shapes[i]);
+            }
+            size_t output_rank = m_body->get_output_partial_shape(0).rank().get_length();
+            if (output_rank > masterShape.size()) {
+                masterShape.insert(masterShape.begin(), output_rank - masterShape.size(), 1);
+            }
+            return {{masterShape}, ShapeInferStatus::success};
+        } else {
+            std::vector<VectorDims> outputDims;
+            std::vector<ov::Shape> new_shapes;
+            for (const auto& s : input_shapes)
+                new_shapes.emplace_back(s);
+            auto& params = m_body->get_parameters();
+            if (params.size() != input_shapes.size()) {
+                IE_THROW() << "Got invalid number of input shapes to reshape subgraph body";
+            }
+            for (size_t i = 0; i < params.size(); ++i) {
+                params[i]->set_partial_shape(new_shapes[i]);
+            }
+            m_body->validate_nodes_and_infer_types();
+            for (const auto& res : m_body->get_results()) {
+                auto& pshape = res->get_input_partial_shape(0);
+                if (!pshape.is_static()) {
+                    IE_THROW() << "Subgraph inferred dynamic output shape during reshape with static inputs";
+                }
+                outputDims.emplace_back(pshape.get_shape());
+            }
+
+            return {outputDims, ShapeInferStatus::success};
+        }
     }
 
     port_mask_t get_port_mask() const override {
         return EMPTY_PORT_MASK;
     }
 
 private:
-    Snippet* m_node;
+    std::shared_ptr<ov::Model> m_body;
 };
 
 class SnippetShapeInferFactory : public ShapeInferFactory {
 public:
-    SnippetShapeInferFactory(Snippet* node) : m_node(node) {}
+    SnippetShapeInferFactory(const std::shared_ptr<ov::Node>& op) {
+        auto subgraph = ov::as_type_ptr<snippets::op::Subgraph>(op);
+        snippet_body = subgraph->body_ptr()->clone();
```

Comment:
Why do we need to clone body function here? Can't we just copy the `body_ptr()` return value?

Reply:
We may have to do it, if some operations from snippets opset break shape propagation. If this is the case, could you identify such ops, so we can fix their `validate_and_infer_types`?

Reply:
The reason of a clone() here is that body_ptr() is shared across streams. Different streams should have its own shape infer with possibly different input shapes, should not impact each other. Am I right?

Reply:
Yes, that's correct. Previously we solved this problem by making a deep copy in the Snippet node constructor.
Now I see that you moved the deep copy from constructor to SnippetShapeInferFactory, so we still have one copy per node.
That's all right as well, so let's leave as is.